modules/classloaders/hdfs-class-loader/src/main/java/org/apache/accumulo/classloader/hdfs/HDFSContextClassLoaderFactory.java

```diff
@@ -0,0 +1,402 @@
+      Context context = new Gson().fromJson(reader, Context.class);
+      urls = new URL[context.getJars().size()];
+      int i = 0;
+      if (HDFSManifestFile.getParent() != null) {
+        for (var jar : context.getJars()) {
+          // download to temp dir
+          var HDFSJarPath = new Path(HDFSManifestFile.getParent(), jar.getJarName());
+          var localTempJarPath = localTempContextDir.resolve(jar.getJarName());
+          hdfs.copyToLocalFile(HDFSJarPath, new Path(localTempJarPath.toUri()));
+          LOG.info("Copied from {} to {}", HDFSJarPath, localTempJarPath);
+          // verify downloaded jar checksum matches what is in the manifest file
+          var computedChecksumLocalJar = checksumLocalFile(localTempJarPath);
+          if (!computedChecksumLocalJar.equals(jar.getChecksum())) {
+            throw new ContextClassLoaderException(contextName,
+                new IllegalStateException(String.format(
+                    "checksum: %s of downloaded jar: %s (downloaded from %s to %s) did not match "
+                        + "checksum present: %s in manifest file: %s. Consider retrying and/or "
+                        + "updating the checksum for this jar in the manifest file",
+                    computedChecksumLocalJar, jar.getJarName(), HDFSJarPath, localTempJarPath,
+                    jar.getChecksum(), HDFSManifestFile)));
+          }
+          // use final (non-temp) jar path for the URL as that is what will exist if this op
+          // completes
+          var localFinalJarPath = localFinalContextDir.resolve(jar.getJarName());
+          urls[i++] = localFinalJarPath.toUri().toURL();
+        }
+      } else {
+        throw new ContextClassLoaderException(contextName);
+      }
+    }
+
+    return new TempState(contextName, localTempContextDir, localFinalContextDir,
+        HDFSManifestFileChecksum, urls);
+  }
+
+  private Pair<String,URLClassLoader> createClassLoader(TempState tempState) throws IOException {
+    // if the final path already exists, we are replacing it with temp path, so recursively delete
+    // the final path
+    if (Files.exists(tempState.finalPath)) {
+      FileUtils.deleteDirectory(tempState.finalPath.toFile());
+    }
+    // rename temp path to final path
+    if (tempState.tempPath.toFile().renameTo(tempState.finalPath.toFile())) {
+      LOG.info("Renamed {} to {}", tempState.tempPath, tempState.finalPath);
+    }
+    var classLoader = new URLClassLoader(tempState.urls, ClassLoader.getSystemClassLoader());
+    return new Pair<>(tempState.checksum, classLoader);
+  }
+
+  public String checksumLocalFile(java.nio.file.Path path) throws IOException {
+    try (var in = FileUtils.openInputStream(path.toFile())) {
+      return checksum(in.readAllBytes());
+    }
+  }
+
+  public static String checksum(byte[] fileBytes) {
+    try {
+      StringBuilder checksum = new StringBuilder();
+      for (byte b : MessageDigest.getInstance(HASH_ALG).digest(fileBytes)) {
```

Comment:
A micro-optimization but thought I'd mention since we are potentially calling this method a lot for large contexts. The `getInstance()` is a service provider interface lookup which isn't the cheapest thing. An alternative I saw proposed was to get an instance when we initialize the Factory and then have this method call `instance.clone()` for a faster way to get a fresh one.